dojo_plugin/api/v1/discord.py

```diff
@@ -56,5 +65,129 @@ def get(self, discord_id):
                 "reference_id": dojo_challenge.reference_id,
             }
         }
-
         return {"success": True, "activity": activity}
+
+
+def get_user_activity_prop(discord_id, activity, start=None, end=None):
+    user = DiscordUsers.query.filter_by(discord_id=discord_id).first()
+
+    if activity is DiscordUserActivity.ActivityType.thanks:
+        prop_name = "thanks"
+        count = user.thanks_count(start, end) if user else 0
+    elif activity is DiscordUserActivity.ActivityType.memes:
+        prop_name = "memes"
+        count = user.memes_count(start, end) if user else 0
+
+    return {"success": True, prop_name: count}
+
+def get_user_activity(self, discord_id, activity, request):
+    authorization = request.headers.get("Authorization")
+    res, code = auth_check(authorization)
+    if res:
+        return res, code
+
+    start_stamp = request.args.get("start")
+    end_stamp = request.args.get("end")
+    start = None
+    end = None
+
+    if start_stamp:
+        try:
+            start = datetime.fromisoformat(start_stamp)
+        except:
+            return {"success": False, "error": "invalid start format"}, 400
+    if end_stamp:
+        try:
+            end = datetime.fromisoformat(start_stamp)
+        except:
+            return {"success": False, "error": "invalid end format"}, 400
+
+    user = DiscordUsers.query.filter_by(discord_id=discord_id).first()
+
+    return get_user_activity_prop(discord_id, activity, start, end)
+
+def post_user_activity(discord_id, activity, request):
+    authorization = request.headers.get("Authorization")
+    res, code = auth_check(authorization)
+    if res:
+        return res, code
+
+    #data = request.get_json()
+    try:
+        data = json.loads(request.get_data())
```

Comment:
This has to be fixed. This is now how you bring json data in.

@bypass_csrf_protection
https://github.com/CTFd/CTFd/blob/c4b83305624a0f5f39700bb54b1e8711c90446f1/CTFd/plugins/__init__.py#L158
https://github.com/CTFd/CTFd/blob/master/CTFd/utils/initialization/__init__.py#L315

Should look just like this:
https://github.com/pwncollege/dojo/blob/5b4ee1db66557f3321da8122e8db9912d3a06cb2/dojo_plugin/pages/dojo.py#L135-L138

Please let me know if you are unable to get this to work.

Reply:
This does not work.

I believe it has to do with discord.py relying on `flask_restx` instead of flask.

Reply:
https://github.com/pwncollege/dojo/commit/0dc67d2732055d1bbae650783455286b83dd8937

This commit should have resolved your issue. Don't worry about specifying `@bypass_csrf_protection`, CTFd automatically bypasses that logic for requests with an `Authorization` header anyways.

Reply:
And yet.. it still doesn't work.

```
  headers = {"Authorization": f"Bearer {DISCORD_CLIENT_SECRET}"}
        data = {"source_user_id": payload.user_id,
                "message_id": payload.message_id,
                "channel_id": payload.channel_id,
                "guild_id": payload.guild_id,
                "message_timestamp": message.created_at.isoformat(),
                }

        async with aiohttp.ClientSession() as session:
            response = await session.post(api_url, headers=headers, json=data)
            data = await response.json()
            print(data)
```

```
<@281565385139159040> liked <@281565385139159040>'s meme
{'message': "The server could not verify that you are authorized to access the URL requested. You either supplied the wrong credentials (e.g. a bad password), or your browser doesn't understand how to supply the credentials required."}
```

Reply:
Nvm.. I see that commit is in master...